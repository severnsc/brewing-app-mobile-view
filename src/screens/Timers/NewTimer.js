import React from "react";
import PropTypes from "prop-types";
import { View } from "react-native";
import { TimerInput, TextInput, Subtitle, Button } from "../../components";

const NewTimer = ({
  onTimerNameChange,
  onTimerDurationChange,
  onAlertMessageChange,
  onAlertTimeChange,
  onPress
}) => (
  <View>
    <TextInput autoFocus={true} onChange={onTimerNameChange} />
    <TimerInput
      placeholder="Timer Duration (HH:MM:SS)"
      onChange={onTimerDurationChange}
    />
    <Subtitle value="Alerts" />
    <TextInput placeholder="Alert Message" onChange={onAlertMessageChange} />
    <TimerInput
      placeholder="Alert Time (HH:MM:SS)"
      onChange={onAlertTimeChange}
    />
    <Button circle={true} success={true} value="+" onPress={onPress} />
  </View>
);

NewTimer.propTypes = {
  onTimerNameChange: PropTypes.func.isRequired,
  onTimerDurationChange: PropTypes.func.isRequired,
  onAlertMessageChange: PropTypes.func.isRequired,
  onAlertTimeChange: PropTypes.func.isRequired,
  onPress: PropTypes.func.isRequired
};

NewTimer.defaultProps = {
  onTimerNameChange: () => {},
  onTimerDurationChange: () => {},
  onAlertMessageChange: () => {},
  onAlertTimeChange: () => {},
  onPress: () => {}
};

export default NewTimer;
