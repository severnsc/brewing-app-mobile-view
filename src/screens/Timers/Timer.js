import React, { useState } from "react";
import PropTypes from "prop-types";
import { View, FlatList } from "react-native";
import {
  Title,
  Clock,
  Button,
  Icon,
  Subtitle,
  Text,
  TextInput,
  TimerInput,
  Swipable
} from "../../components";

const icon = <Icon name="md-create" />;
const separator = () => (
  <View
    style={{
      borderBottomWidth: 1,
      borderBottomColor: "#000"
    }}
  />
);

const ReadMode = ({ title, duration, alerts, toggleEditing }) => (
  <View>
    <Title value={title} />
    <Clock size="sm" ms={duration} />
    <Button
      circle={true}
      primary={true}
      value={icon}
      onPress={toggleEditing}
      testID="editButton"
    />
    <Button danger={true} value="Delete timer" />
    <Subtitle value={alerts.length + " alerts"} />
    <FlatList
      data={alerts}
      renderItem={item => (
        <React.Fragment>
          <Text>{item.message}</Text>
          <Clock size="md" ms={item.activationTime} />
        </React.Fragment>
      )}
      ItemSeparatorComponent={separator}
    />
    <Button value="Activate Timer" success={true} />
  </View>
);

const EditMode = ({
  title,
  duration,
  alerts,
  toggleEditing,
  onAlertMessageChange
}) => (
  <View>
    <TextInput value={title} />
    <TimerInput value={duration} />
    <Subtitle value={alerts.length + " alerts"} />
    <FlatList
      data={alerts}
      renderItem={item => {
        const deleteButton = <Button danger={true} value="Delete" />;
        return (
          <Swipable swipeLeftComponent={deleteButton}>
            <View>
              <TextInput value={item.message} onChange={onAlertMessageChange} />
              <TimerInput value={item.activationTime} />
            </View>
          </Swipable>
        );
      }}
      ItemSeparatorComponent={separator}
    />
    <Button
      circle={true}
      success={true}
      value={<Icon name="ios-add-circle" />}
    />
  </View>
);

const Timer = ({ title, duration, alerts, onAlertMessageChange }) => {
  const [isEditing, setIsEditing] = useState(false);
  const toggleEditing = () => setIsEditing(!isEditing);
  return isEditing ? (
    <EditMode
      title={title}
      duration={duration}
      alerts={alerts}
      onAlertMessageChange={onAlertMessageChange}
    />
  ) : (
    <ReadMode
      title={title}
      duration={duration}
      alerts={alerts}
      toggleEditing={toggleEditing}
    />
  );
};

Timer.propTypes = {
  title: PropTypes.string,
  duration: PropTypes.number,
  alerts: PropTypes.array.isRequired,
  onAlertMessageChange: PropTypes.func.isRequired
};

Timer.defaultProps = {
  title: "",
  duration: 0,
  alerts: [],
  onAlertMessageChange: () => {}
};

export default Timer;
