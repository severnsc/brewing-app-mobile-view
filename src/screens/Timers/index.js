import Timers from "./Timers";
import NewTimer from "./NewTimer";
import Timer from "./Timer";

export { Timers, NewTimer, Timer };
