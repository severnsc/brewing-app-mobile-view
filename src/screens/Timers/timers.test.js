import React from "react";
import { shallow } from "enzyme";
import Timers from ".";

const timerObjects = [
  {
    id: "1",
    name: "Timer 1",
    duration: 25000,
    alerts: [
      {
        id: "1"
      },
      {
        id: "2"
      }
    ]
  },
  {
    id: "2",
    name: "Timer 2",
    duration: 129999,
    alerts: [
      {
        id: "1"
      },
      {
        id: "2"
      },
      {
        id: "3"
      }
    ]
  }
];

describe("Timers screen", () => {
  it("should return timers title", () => {
    const timers = shallow(<Timers />);
    const title = timers.find("Title");
    expect(title.prop("value")).toBe("Timers");
  });
  it("should return a FlatList", () => {
    const timers = shallow(<Timers />);
    const list = timers.find("FlatList");
    expect(list).toHaveLength(1);
  });
  it("should return a card in the list for each timer object", () => {
    const timers = shallow(<Timers timers={timerObjects} />);
    const list = timers.find("FlatList");
    const cards = list.props().data.map(list.props().renderItem);
    expect(cards).toHaveLength(2);
  });
  it("should format the card correctly", () => {
    const timers = shallow(<Timers timers={timerObjects} />);
    const list = timers.find("FlatList");
    const cards = list.props().data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
});
