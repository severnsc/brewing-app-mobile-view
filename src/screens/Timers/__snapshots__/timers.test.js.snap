// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Timer screen matches snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "borderBottomColor": "#37c172",
        "borderBottomWidth": 1,
        "color": "#37c172",
      }
    }
  >
    <TextInput
      autoFocus={false}
      keyboardType="default"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
    />
  </View>
  <View>
    <RCTInputAccessoryView
      backgroundColor="#f1f3f5"
      style={
        Array [
          undefined,
          Object {
            "position": "absolute",
          },
        ]
      }
    >
      <Text
        accessibilityLabel="Done"
        onPress={[Function]}
        style={
          Object {
            "color": "#3183c8",
            "padding": 10,
            "textAlign": "right",
          }
        }
      >
        Done
      </Text>
    </RCTInputAccessoryView>
    <View
      style={
        Object {
          "borderBottomColor": undefined,
          "borderBottomWidth": 1,
        }
      }
    >
      <TextInput
        autoFocus={false}
        keyboardType="numeric"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Timer Duration (HH:MM:SS)"
        value=""
      />
    </View>
  </View>
  <Text
    accessibilityLabel="Alerts"
    style={
      Object {
        "color": undefined,
        "fontSize": 24,
      }
    }
  >
    Alerts
  </Text>
  <View
    style={
      Object {
        "borderBottomColor": undefined,
        "borderBottomWidth": 1,
      }
    }
  >
    <TextInput
      autoFocus={false}
      keyboardType="default"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Alert Message"
    />
  </View>
  <View>
    <RCTInputAccessoryView
      backgroundColor="#f1f3f5"
      style={
        Array [
          undefined,
          Object {
            "position": "absolute",
          },
        ]
      }
    >
      <Text
        accessibilityLabel="Done"
        onPress={[Function]}
        style={
          Object {
            "color": "#3183c8",
            "padding": 10,
            "textAlign": "right",
          }
        }
      >
        Done
      </Text>
    </RCTInputAccessoryView>
    <View
      style={
        Object {
          "borderBottomColor": undefined,
          "borderBottomWidth": 1,
        }
      }
    >
      <TextInput
        autoFocus={false}
        keyboardType="numeric"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Alert Time (HH:MM:SS)"
        value=""
      />
    </View>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#37c172",
        "borderColor": "#e9ecee",
        "borderRadius": 100,
        "borderWidth": 1,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
        "width": 50,
      }
    }
  >
    <Text
      accessibilityLabel="+"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
        }
      }
    >
      +
    </Text>
  </View>
</View>
`;

exports[`Timer screen editing mode returns a FlatList of alert inputs 1`] = `
Array [
  <Swipable
    swipeLeftComponent={
      <Button
        danger={true}
        value="Delete"
      />
    }
  >
    <View>
      <TextInput
        keyboardType="default"
        onChange={[MockFunction]}
        value="Add bittering hops"
      />
      <TimerInput
        value={3600000}
      />
    </View>
  </Swipable>,
  <Swipable
    swipeLeftComponent={
      <Button
        danger={true}
        value="Delete"
      />
    }
  >
    <View>
      <TextInput
        keyboardType="default"
        onChange={[MockFunction]}
        value="Sanitize chiller"
      />
      <TimerInput
        value={300000}
      />
    </View>
  </Swipable>,
  <Swipable
    swipeLeftComponent={
      <Button
        danger={true}
        value="Delete"
      />
    }
  >
    <View>
      <TextInput
        keyboardType="default"
        onChange={[MockFunction]}
        value="Add whirlpool hops"
      />
      <TimerInput
        value={0}
      />
    </View>
  </Swipable>,
]
`;

exports[`Timer screen returns a FlatList of alerts 1`] = `
Array [
  <React.Fragment>
    <Text>
      Add bittering hops
    </Text>
    <Clock
      ms={3600000}
      size="md"
    />
  </React.Fragment>,
  <React.Fragment>
    <Text>
      Sanitize chiller
    </Text>
    <Clock
      ms={300000}
      size="md"
    />
  </React.Fragment>,
  <React.Fragment>
    <Text>
      Add whirlpool hops
    </Text>
    <Clock
      ms={0}
      size="md"
    />
  </React.Fragment>,
]
`;

exports[`Timers screen should format the card correctly 1`] = `
Array [
  <Card
    lowerLeft={
      <Clock
        ms={25000}
        size="sm"
      />
    }
    lowerRight={
      <Text
        value="2 alerts"
      />
    }
    upperLeft={
      <Subtitle
        value="Timer 1"
      />
    }
  />,
  <Card
    lowerLeft={
      <Clock
        ms={129999}
        size="sm"
      />
    }
    lowerRight={
      <Text
        value="3 alerts"
      />
    }
    upperLeft={
      <Subtitle
        value="Timer 2"
      />
    }
  />,
]
`;
